Update DEVELOPING.MD for readability (#913)
mapbox/mapbox-maps-ios: DEVELOPING.md

3. uses [Swift Package Manager](https://github.com/apple/swift-package-manager) to manage dependencies for development. Cocapods is supported for *consuming* the SDK.

4. uses [secret-shield](https://github.com/mapbox/secret-shield) to help block secrets such as access tokens from being exposed. Setup `secret-shield` by running:
```
npm install -g @mapbox/secret-shield
scripts/install-pre-commit/install-pre-commit.sh
```

> 5. uses CircleCI and AWS Device Farm for continuous integration.

Reviewer: Should be updated to refer to Firebase Test Lab now (check for other references as well)

### Accessing the Maps SDK's source

Clone the git repository:

```
git clone git@github.com:mapbox/mapbox-maps-ios.git && cd mapbox-maps-ios
```

### Building the Maps SDK

> Pre-requisite: Valid `.netrc` file located on your machine at `~/.netrc`. Please visit https://docs.mapbox.com/ios/maps/overview for additional information about setting up your `.netrc` file.

Reviewer: Let's explain why this is necessary: it allows Swift Package Manager to download binary dependencies from Mapbox SDK Registry.

Open `Package.swift` and build the MapboxMaps target.

### Update dependency versions

You must remember to update the Package.swift file in https://github.com/mapbox/mapbox-maps-ios/blob/main/Package.swift, not the App.xcworkspace, using the following commands:

```
xed Package.swift
//update versions in there
swift package update

> ```

Reviewer: Updating dependencies has more steps:

* Update the dependency versions in Package.swift
* Open Package.swift in Xcode and resolve dependencies. This updates Package.resolved. Close that Xcode workspace.
* Open Apps/Apps.xcworkspace in Xcode and resolve dependencies. This updates a different Package.resolved file in the Apps.workspace bundle. Close that Xcode workspace.
* Update the dependency versions in MapboxMaps.podspec
* Update the dependency versions in `scripts/release/packager/versions.json`

Commit the changes to all 5 files and open a PR.

### Running the Debug App

> In order to use the debug app, open Apps/Apps.xcworkspace and build and run the `DebugApp` target.

Reviewer: ```suggestion
In order to use the debug app, open Apps/Apps.xcworkspace and build and run the `DebugApp` scheme.
```

You must provide a Mapbox access token to display Mapbox-hosted maps in the `DebugApp` test application. Add your Mapbox token to the `MBXAccessToken` key in `DebugApp/DebugApp/Info.plist`. Alternatively, create a [file at `~/mapbox`](https://docs.mapbox.com/help/troubleshooting/private-access-token-android-and-ios/#ios) containing your access token.

### CircleCI

> CircleCI's [YAML](https://en.wikipedia.org/wiki/YAML) format can be picky. You can run `make validate` to ensure you've formatted the changes to the config file correctly. This make target will install the [CircleCI command line interface](https://circleci.com/docs/2.0/local-cli/) to validate `.circleci/config.yml`, and run `circleci config validate`

Reviewer: I think we can remove the criticism of CircleCI's YAML format 🙃 

The advice about running `make validate` is nice though.

# Testing

## Unit Testing on devices

To run device tests there are few options:

1. Run via Xcode:

  - Install [xcodegen](https://github.com/yonaskolb/XcodeGen).
  - Run `$ xcodegen` in the root of the repo.
  - Open the resulting `MapboxMaps.xcodeproj`.
  - Test (Cmd-U) the `MapboxTestHost` scheme.

2. Trigger via CI by adding `[run device tests]` to a git commit message. These tests also run "nightly".
(Examples can also be run as tests on CI by adding `[run app device tests]`.)

3. Trigger tests on AWS Device Farm from the command line. Note that this requires certain environment variables to be set; please see the makefile for these.
`make test-with-device-farm SCHEME=MapboxTestHost APP_NAME=MapboxTestHost`.

4. Trigger tests on a local device (connected by USB) using the same setup as Device Farm testing by running:

> `make local-test-with-device-farm-ipa SCHEME=MapboxTestHost CONFIGURATION=Release ENABLE_CODE_SIGNING=1`

Reviewer: I'm not sure any of these are supported anymore. @OdNairy has been working on some improvements here. I know for 2, we know trigger device tests in the CircleCI web app.

Reviewer 2: We do not support that makefile rules anymore

## Integration Tests

> Integration tests typically test the integration between components. As such most will require a `MapView`. These can be run locally and on AWS Device Farm.

Reviewer: ```suggestion
Integration tests typically require a Metal device, so these tests can only run locally and on Firebase Test Lab. They are skipped when running on CircleCI inside of a VM.
```

> Since these tests require a map view, they also need a `UIWindow` and Metal rendering to work. For both testing scenarios, if there's no valid Metal device the test will be skipped (i.e. not failure/success).

Reviewer: Can be deleted

### What happens if I run from `MapboxTestHost`?

There **is** a host application, so `MapViewIntegrationTestCase` fetches the existing window and view controller before adding the MapView to it.

- These tests can be run on devices.
- Tests will be skipped on simulators < iOS 13 because there's no valid Metal device. (iOS 13+ has a simulated Metal device.)

### What happens if I run from `MapboxMaps`

There is **no host application**, so `MapViewIntegrationTestCase` creates its own `UIWindow` and root view controller, before adding the MapView to it.

- These tests cannot be run on devices.
- Tests will be skipped on simulators < iOS 13 because there's no valid Metal device.

> - Tests will be skipped entirely on CircleCI because of the VM (no Metal). **So for CI purposes, integration tests should be run on AWS Device Farm via `MapboxTestHost`**

Reviewer: We should add a note somewhere about *when* to run integration tests in CI since it's opt-in. @OdNairy @evil159 any opinions of when this should happen? Should we require it for PR merge, but agree to typically only run them after code review? Or just keep them optional?

Reviewer: They run post-merge either way.

As you can see the most useful case is to run within from the host application. Be aware that this has the potential for side-effects, since the application is not restarted for each test.

### Base classes

- `IntegrationTestCase` is a base class that fetches the access token either from the Info.plist or from `UserDefaults` (which will override). The Info.plist in `MapboxMaps` has been set up to use the variable `MAPBOX_ACCESS_TOKEN`. **Please add a `developer.xcconfig` in the root of the repo, that looks like:**

    ```
    MAPBOX_ACCESS_TOKEN = pk.myaccesstoken
    ```

>     The CircleCI config file has been set up to create this.

Reviewer: @OdNairy is this still true?

> - `MapViewIntegrationTestCase` subclasses the above, and sets the `MapView`, and is the `MBXMapViewDelegate`. Closures are used to expose these callbacks to subclasses. Please don't add tests to this class.

Reviewer: `MBXMapViewDelegate` isn't a thing anymore

- `ExampleIntegrationTest` is an example of using the above, that sets a style then waits for the load to finish.

## Application Testing

### Making an Example

> * Check out this [doc](https://github.com/mapbox/mapbox-maps-ios/blob/main/Apps/Examples/README.md) to get more information about adding examples to our project.### Setting up a development environment

Reviewer: ```suggestion
* Check out this [doc](https://github.com/mapbox/mapbox-maps-ios/blob/main/Apps/Examples/README.md) to get more information about adding examples to our project.

### Setting up a development environment
```

This project runs on Apple's Metal rendering framework. Prerequisites for running the test app in simulators:

- macOS 10.15 or later

> - Xcode 12.2+

Reviewer: ```suggestion
- Xcode 12.5.1+
```

This project:

1. requires [a valid ~/.netrc file](https://docs.mapbox.com/ios/maps/guides/install/#configure-credentials) to fetch dependencies.

2. uses [SwiftLint](https://github.com/realm/SwiftLint) to enforce a style guide. Install it with `brew install swiftlint` and check for lint by running `swiftlint` in the root of the repository.

3. uses [Swift Package Manager](https://github.com/apple/swift-package-manager) to manage dependencies for development. Cocapods is supported for *consuming* the SDK.

4. uses [secret-shield](https://github.com/mapbox/secret-shield) to help block secrets such as access tokens from being exposed. Setup `secret-shield` by running:
```
npm install -g @mapbox/secret-shield
scripts/install-pre-commit/install-pre-commit.sh

> ```

Reviewer: This feels redundant with a very similar section above. Is there a reason to duplicate it?

  - Open the resulting `MapboxMaps.xcodeproj`.
  - Test (Cmd-U) the `MapboxTestHost` scheme.

2. Trigger via CI by adding `[run device tests]` to a git commit message. These tests also run "nightly".
(Examples can also be run as tests on CI by adding `[run app device tests]`.)

3. Trigger tests on AWS Device Farm from the command line. Note that this requires certain environment variables to be set; please see the makefile for these.
`make test-with-device-farm SCHEME=MapboxTestHost APP_NAME=MapboxTestHost`

4. Trigger tests on a local device (connected by USB) using the same setup as Device Farm testing by running:

> `make local-test-with-device-farm-ipa SCHEME=MapboxTestHost CONFIGURATION=Release ENABLE_CODE_SIGNING=1`

Reviewer: Also feels redundant…

### Making an Example

> * Check out this [doc](https://github.com/mapbox/mapbox-maps-ios/blob/main/Apps/Examples/README.md) to get more information about adding examples to our project.

Reviewer: Redundant?
